src/Storages/Cache/ExternalDataSourceCache.cpp

```diff
@@ -57,8 +57,15 @@ LocalFileHolder::~LocalFileHolder()
 {
     if (original_readbuffer)
     {
-        assert_cast<SeekableReadBuffer *>(original_readbuffer.get())->seek(0, SEEK_SET);
-        file_cache_controller->value().startBackgroundDownload(std::move(original_readbuffer), *thread_pool);
+        try
+        {
+            assert_cast<SeekableReadBuffer *>(original_readbuffer.get())->seek(0, SEEK_SET);
+            file_cache_controller->value().startBackgroundDownload(std::move(original_readbuffer), *thread_pool);
```

Comment:
Technically, the fix is ok. But could you clarify what's the point of starting background download in `LocalFileHolder` destructor?

cc @lgbo-ustc

Reply:
@devcrafter I also have no idea why. This was done because we were running some tests on object storage and then ClickHouse crashed; tracing the coredump led us here.
May be @lgbo-ustc will know more.

Reply:
We reuse the readbuffer here, so we cannot use the read buffer for current query process and download at the same time. The file can only be download after its first access finished, `LocalFileHolder` is the place we know the query process finished.

Reply:
> @devcrafter I also have no idea why. This was done because we were running some tests on object storage and then ClickHouse crashed; tracing the coredump led us here. May be @lgbo-ustc will know more.

What is the exception ?

Reply:
It'd be nice to have some details regarding the particular exception, but in general, it seems clear that `startBackgroundDownload()` (for example, constructor `WriteBufferFromFile` can throw)